「Contextを使ってグローバルなstateを管理する」の課題

ham-cap opened the pull request:
課題を提出いたします。
よろしくお願いいたします🙇‍♂️

ham-cap committed:
グローバルstateとしてIsLoggedInを設定・Form側での参照テスト

ham-cap committed:
新規フォームを開く+ボタンの表示/非表示をisLoggedInの値によって切り替えるように変更

ham-cap committed:
ログイン/ログアウトボタンを追加

ham-cap committed:
ログインボタンによる編集、登録、削除ボタンの表示切り替えを実装

ham-cap committed:
Contextの作成部分を個別のファイルへ切り出した

ham-cap committed:
isLoggedInとその更新関数を返すカスタムフックを作成し、App.jsxに導入

ham-cap committed:
ログインにまつわるContextをIsLoggedInProviderに切り出してカプセル化した

ham-cap committed:
前回レビュー時の修正事項を反映

【useMemos.jsx】
・JSXを使用していないためuseMemos.jsxの拡張子を.jsに変更した。

【MemosSection】
・LocalStorageから取得したデータを格納する変数名をmemosJsonとしていたが、データの取得時点ではJSON形式ではないため修正した。
・LocalStorageからデータを取得するための記述が重複していたため、不要なほうを削除した。
・isVisibleという名称だと何を可視化するか分かりにくいため、フォームの表示/非表示切り替えに使用することを連想できるよう変更した。

ham-cap committed:
不要なコメントを削除

ham-cap committed:
Prettierの修正を反映

ham-cap committed:
ESLintの修正を反映

ham-cap committed:
コンフリクト修正

yuuu commented on src/IsLoggedInProvider.jsx:
```diff
@@ -0,0 +1,25 @@
+}
+
+export function useIsLoggedIn () {
+  const { isLoggedIn, setIsLoggedIn } = useContext(IsLoggedInContext)
```
ここは以下のように書けますね。

```javascript
export function useIsLoggedIn () {
  return useContext(IsLoggedInContext)
}
```

アロー関数を使うとさらにスッキリ書けます。

```javascript
export const useIsLoggedIn = () => useContext(IsLoggedInContext)
```

yuuu replied:
`setIsLoggedIn` をそのままコンポーネントに渡すと、利用する側としては具体的過ぎて扱いづらいですね。
例えば、 `login` `logout` といった関数でラッピングしてそれを返すようにすると、抽象度が上がることで利用する側は何ができるのか直感的に理解できますし、引数の指定も不要となって使いやすくなります 👍

yuuu commented on src/IsLoggedInProvider.jsx:
```diff
@@ -0,0 +1,25 @@
+
+const IsLoggedInContext = createContext()
+
+function IsLoggedInProvider (props) {
```
`login` `logout` のような関数を提供するのであれば、 `IsLoggedInProvider` という名前も具体的過ぎるので、もう少し抽象度の高い名前にしたほうが良いですね。

ham-cap committed:
IsLoggedInProviderを修正

・他のコンポーネントがContextを利用する際に挙動などを直感的に理解しやすいように、IsLoggedInProviderが提供するsetIsLoggedIn関数をswitchLogInStatus関数でラップした状態で提供することとした。
また、それに合わせてコンポーネント自体とカスタムフックの名称も変更した。

ham-cap commented:
アドバイスをもとに改善してみました！
よろしくお願いいたします🙇‍♂️

yuuu commented on src/LogInStatusProvider.jsx:
```diff
@@ -0,0 +1,24 @@
+  )
+}
+
+export const useLogInStatus = () => useContext(IsLoggedInContext)
```
`useIsLoggedIn` → `useLogInStatus` にリネームいただきましたが、あまり抽象度が変わっていないと感じました。
「ログイン状態」というよりは、「ログイン」「ログアウト」「ログインしているか確認」といった機能を提供するHooksなので、それらをまとめてなんと呼ぶか？を考えて名前を付けてみてください。Providerについても同様です。

yuuu commented on src/LogInStatusProvider.jsx:
```diff
@@ -0,0 +1,24 @@
+function LogInStatusProvider (props) {
+  const [isLoggedIn, setIsLoggedIn] = useState(false)
+
+  const switchLogInStatus = () => setIsLoggedIn(!isLoggedIn)
```
ここも、利用する側は `Status` という具体を意識する必要はないのでメソッドは分かれてしまいますがシンプルに `LogIn` `Logout` で良いのでは？と感じました。

ham-cap committed:
LogInStatusProviderを修正

・カスタムフックとそれを提供するコンポーネントの名称をより抽象度の高いものに変更
・利用側にとってはログインとログアウトを司る関数であることが伝わればいいため、switchLogInStatusをlogInとlogOutの二つに分割した

ham-cap commented:
再度修正してみました！
ご確認お願いいたします🙏
